src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,469 @@
+    function transferFrom(address, address, uint256) external returns (bool);
+    function scaledBalanceOf(address) external view returns (uint256);
+    function decimals() external view returns (uint8);
+}
+
+interface ChainlogLike {
+    function getAddress(bytes32) external view returns (address);
+}
+
+interface DaiJoinLike {
+    function dai() external view returns (address);
+}
+
+interface VatLike {
+    function hope(address) external;
+    function ilks(bytes32) external view returns (uint256, uint256, uint256, uint256, uint256);
+    function urns(bytes32, address) external view returns (uint256, uint256);
+    function gem(bytes32, address) external view returns (uint256);
+    function live() external view returns (uint256);
+    function slip(bytes32, address, int256) external;
+    function move(address, address, uint256) external;
+    function frob(bytes32, address, address, address, int256, int256) external;
+    function grab(bytes32, address, address, address, int256, int256) external;
+    function fork(bytes32, address, address, int256, int256) external;
+    function suck(address, address, uint256) external;
+}
+
+interface EndLike {
+    function debt() external view returns (uint256);
+    function skim(bytes32, address) external;
+}
+
+interface DssDirectDepositJoinLike {
+    function getMaxBar() external view returns (uint256);
+    function validTarget() external view returns (bool);
+    function calcSupplies(uint256, uint256) external view returns (uint256, uint256);
+    function supply(uint256) external;
+    function withdraw(uint256) external;
+    function collect(address[] memory, uint256, address) external returns (uint256);
+    function gemBalanceOf() external view returns(uint256);
+    function getNormalizedBalanceOf() external view returns(uint256);
+    function getNormalizedAmount(uint256) external view returns(uint256);
+    function cage() external;
+}
+
+contract DssDirectDepositHub {
+
+    // --- Auth ---
+    mapping (address => uint) public wards;
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+
+        emit Deny(usr);
+    }
+    modifier auth {
+        require(wards[msg.sender] == 1, "DssDirectDepositHub/not-authorized");
+        _;
+    }
+
+    struct D3M {
+        DssDirectDepositJoinLike join;
+        TokenLike                gem;
+        uint256                  tau; // Time until you can write off the debt [sec]
+        uint256                  bar; // Target Interest Rate [ray]
+        uint256                  culled;
+        address                  king; // Who gets the rewards
+        uint256                  tic; // Timestamp when the join is caged
+    }
+
+    ChainlogLike public immutable chainlog;
+    VatLike public immutable vat;
+    mapping (bytes32 => D3M) public d3ms;
```

Comment:
My first instict was that this should be called `ilks`.

```suggestion
    mapping (bytes32 => D3M) public ilks;
```

Reply:
I had it that way originally the changed it for some reason... but i agree since i am enforcing that they are the same

Reply:
updated in https://github.com/makerdao/dss-direct-deposit/pull/31/commits/769fd8c9d3089bb47af0e50f15f4ce14bceb874b